patterns/behavioural/inner-trait.md

```diff
@@ -0,0 +1,105 @@
+# Inner Trait
+
+## Description
+
+It is possible to define a private trait that implements all the
+methods of a public trait and also includes some private functions. This pattern
+can be used to provide additional functionality to the implementation of a
+public trait while keeping the private methods hidden from the public API.
+
+## Example
+
+This example demonstrate how a public trait `Car` can be implemented and include
+extra private methods using a auxiliary private trait `InnerCar`.
+
+```rust,ignore
+// trait that is public and part of the API
+pub trait Car {
+    fn get_speed(&self) -> f64;
+    fn accelerate(&mut self, duration: f64);
+    fn brake(&mut self, force: f64);
+}
+//not public
+mod inner_lib {
+    // trait that is only accessible to this crate
+    pub(crate) trait InnerCar {
+        fn get_speed(&self) -> f64;
+        //private
+        fn set_speed(&mut self, new_speed: f64);
+        //private
+        fn get_acceleration(&self) -> f64;
+        fn accelerate(&mut self, duration: f64) {
+            self.set_speed(
+                self.get_speed() + (self.get_acceleration() * duration)
+            );
+        }
+        fn brake(&mut self, force: f64) {
+            self.set_speed(
+                self.get_speed() - (force * self.get_acceleration())
+            );
+        }
+    }
+    //Auto implement Car for all InnerCar, by forwarding the Car trait to the
+    //InnerCar implementation
+    impl<T: InnerCar> crate::Car for T {
+        fn get_speed(&self) -> f64 {
+            <Self as InnerCar>::get_speed(self)
+        }
+        fn accelerate(&mut self, duration: f64) {
+            <Self as InnerCar>::accelerate(self, duration)
+        }
+        fn brake(&mut self, force: f64) {
+            <Self as InnerCar>::brake(self, force)
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct Car1(f64);
+//is not necessary to implement `accelerate` and `brake`, as inner_trait can do that.
+impl inner_lib::InnerCar for Car1 {
+    fn get_speed(&self) -> f64 {self.0}
+    fn set_speed(&mut self, new_speed: f64) {self.0 = new_speed}
+    fn get_acceleration(&self) -> f64 {0.10}
+}
+//more Car implementations...
+```
+
+## Motivation
+
+This pattern allows developers to provide additional functionality to the
+implementation of a public trait without exposing that functionality as part of
+the public API. By using a private trait, developers can also improve the
+reusability of their code, since the private functionality can be reused across
+multiple implementations of the public trait.
```

Comment:
Suppose I'm using a trait `Car` from another crate `vehicles`, but for a very specific purpose requiring some additional code. Do you feel that a new trait `InnerCar` private to my own crate would fall under this pattern? If yes, this would be an alternative motivation, wouldn't it?

Reply:
There is another layer on top of this is to prevent breaking changes, since we can now limit functionality without exposing public API.